File: 01-machine/07-inside-a-container/06-memory-the-limit-that-kills-you-without-a-traceback/python/oom.py
"""
7.6 -- Python: there is no heap ceiling, so the OOM killer is your heap limit.

WHAT THIS DEMONSTRATES
  CPython's allocator keeps asking the OS for arenas until the OS stops
  answering, and inside a container the OS stops answering by killing you.
  Nothing in the interpreter reads memory.max. There is no MemoryError to
  catch, no atexit hook, no shutdown log line, no last-gasp metric --
  SIGKILL cannot be caught, blocked or handled, and the process is gone
  between one bytecode and the next.

  This program installs, on purpose, every piece of error handling a
  careful engineer would reach for:

    * try/except MemoryError around the allocation
    * a SIGTERM handler that logs
    * an atexit hook
    * a finally block

  and then allocates until the cgroup kills it. NONE of them run. Reading
  that list of things that did not happen is the point of the exercise --
  the language has the error paths, and the kernel's policy means you never
  reach them.

  Two Python specifics worth knowing while you watch RSS climb:
    * Under Linux's default overcommit, the allocation SUCCEEDS. The charge
      happens when you touch the page, so the failure lands on a memory
      write, arbitrarily far from the allocation.
    * Freed objects go back to pymalloc's pools and arenas, and arenas are
      returned to the OS only when completely empty. RSS is sticky: a burst
      that briefly needed 300MB leaves a process that looks like it needs
      300MB. The --free phase demonstrates that.

WHAT TO LOOK FOR IN THE OUTPUT
  1. The last line printed before the process disappears. There is no
     traceback after it, because there is no exception -- compare that with
     what nodejs/oom.js and java/Oom.java print at the same moment.
  2. `memory.events`' oom_kill counter, read from the shell afterwards. The
     evidence exists; it is just not in your application's logs.
  3. The exit code: 137. Derive it rather than memorise it -- a shell
     reports a signal-terminated process as 128 + signal, and SIGKILL is 9.
  4. In the --free phase, RSS after freeing everything. It does not return
     to where it started.

RUN
    # the real thing: inside a Linux container, with a real cgroup
    docker run --rm --memory=256m -v "$PWD:/w" -w /w python:3.13-slim python oom.py
    echo "exit code: $?"        # 137

    # on this Mac, where there is no cgroup memory controller at all
    python3 oom.py

  On macOS there is no memory.max, no memory.events and no cgroup OOM
  killer. Allocating until something happens would page, swap, and
  eventually annoy you -- a different experiment with a different lesson.
  So on a host with no cgroup limit this program imposes its OWN ceiling
  (--limit-mb, default 512) and stops there, saying clearly that it stopped
  itself and was not killed. That distinction is the whole sub-topic.
"""
from __future__ import annotations

import argparse
import atexit
import os
import platform
import resource
import signal
import sys
import time
from pathlib import Path

HARNESS = Path(__file__).resolve().parents[2] / "00-harness" / "local"
sys.path.insert(0, str(HARNESS))

import cgroup  # noqa: E402

CHUNK_MB = 8


HAVE_CURRENT_RSS = Path("/proc/self/status").exists()


def rss_mb() -> float:
    """Resident set size in MiB -- and NOT the same reading on both platforms.

    /proc/self/status VmRSS is the CURRENT resident set. Darwin has no
    /proc, and getrusage's ru_maxrss is the PEAK, which never goes down.
    Two different questions wearing one function name is exactly the sort of
    thing that produces a confidently wrong memory dashboard, so
    HAVE_CURRENT_RSS is exported and the --free phase below refuses to draw
    a conclusion from a peak.

    (ru_maxrss is also in KiB on Linux and BYTES on macOS. That units
    difference has produced its own share of wrong dashboards.)
    """
    if HAVE_CURRENT_RSS:
        for line in Path("/proc/self/status").read_text().splitlines():
            if line.startswith("VmRSS:"):
                return int(line.split()[1]) / 1024
    peak = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
    return peak / (1024 * 1024) if sys.platform == "darwin" else peak / 1024


def report_and_die(signum, _frame):
    """A SIGTERM handler that logs. It will run for SIGTERM and it will
    never run for SIGKILL, which is the entire difference between a
    graceful shutdown and an OOM kill."""
    print(f"  [signal handler] caught {signal.Signals(signum).name} -- "
          "shutting down cleanly", flush=True)
    raise SystemExit(128 + signum)


@atexit.register
def farewell() -> None:
    """An atexit hook. Runs on a normal exit, on SystemExit, on an uncaught
    exception. Does not run on SIGKILL. If you see this line after an OOM
    kill, you were not OOM-killed."""
    print(f"  [atexit] final RSS {rss_mb():.0f} MiB -- process exiting normally",
          flush=True)


def main() -> None:
    parser = argparse.ArgumentParser(description="7.6 -- the limit that kills without a traceback")
    parser.add_argument("--limit-mb", type=int, default=512,
                        help="self-imposed ceiling when there is no cgroup (default 512)")
    parser.add_argument("--free", action="store_true",
                        help="also demonstrate that RSS does not come back down")
    args = parser.parse_args()

    signal.signal(signal.SIGTERM, report_and_die)

    limit = cgroup.memory_limit()
    high = cgroup.memory_high()
    host = cgroup.host_memory_bytes()
    events_before = cgroup.memory_events()

    print("7.6 -- memory: Python")
    print(f"  interpreter          : CPython {sys.version.split()[0]} on "
          f"{platform.system()} {platform.machine()}")
    print(f"  memory.max           : "
          f"{'no limit / no cgroupfs' if limit is None else f'{limit / 2**20:.0f} MiB'}")
    print(f"  memory.high          : "
          f"{'unset' if high is None else f'{high / 2**20:.0f} MiB'}"
          "   <- the version you can debug, and Compose has no key for it")
    print(f"  /proc/meminfo says   : "
          f"{'?' if host is None else f'{host / 2**20:.0f} MiB'}"
          "   <- the HOST's memory. Not namespaced, so not yours")
    print(f"  Python's heap ceiling: none. There is no such thing.")
    soft, hard = resource.getrlimit(resource.RLIMIT_AS)
    print(f"  RLIMIT_AS            : "
          f"{'unlimited' if soft == resource.RLIM_INFINITY else f'{soft / 2**20:.0f} MiB'}"
          "   <- a DIFFERENT limit, enforced differently")
    print(f"  starting RSS         : {rss_mb():.0f} MiB"
          f"   ({'current, /proc/self/status' if HAVE_CURRENT_RSS else 'PEAK, getrusage -- no /proc here'})")
    print()

    if limit is None:
        print(f"  !! No cgroup memory limit on this host, so nothing can OOM-kill this")
        print(f"  !! process. It will stop ITSELF at {args.limit_mb} MiB and say so.")
        print(f"  !! That is not the experiment -- it is the shape of the experiment")
        print(f"  !! with the lethal part removed. For the real thing:")
        print(f"  !!   docker run --rm --memory=256m -v \"$PWD:/w\" -w /w \\")
        print(f"  !!     python:3.13-slim python oom.py")
        print()
        ceiling_mb = args.limit_mb
    else:
        # Aim past the limit. If the kernel can reclaim page cache fast
        # enough it will, so this has to allocate ANONYMOUS pages and touch
        # them -- reclaimable memory does not get you killed.
        ceiling_mb = int(limit / 2**20 * 1.5)
        print(f"  Allocating toward {ceiling_mb} MiB against a {limit / 2**20:.0f} MiB limit.")
        print("  Every chunk is touched, because under Linux's default overcommit the")
        print("  allocation itself is free -- the cgroup charge lands on the WRITE.")
        print()

    print("  installed, and about to be shown useless:")
    print("    * try/except MemoryError around the allocation")
    print("    * a SIGTERM handler that logs")
    print("    * an atexit hook")
    print("    * a finally block")
    print()

    blocks: list[bytearray] = []
    try:
        while len(blocks) * CHUNK_MB < ceiling_mb:
            # bytearray, not a list of ints: this is a single contiguous
            # allocation whose pages we then touch, which is what a big
            # query result or a JSON serialisation actually looks like.
            block = bytearray(CHUNK_MB * 2**20)
            # Touch every page. One byte per 4 KiB page is enough -- the
            # charge is per page, not per byte.
            for offset in range(0, len(block), 4096):
                block[offset] = 1
            blocks.append(block)

            allocated = len(blocks) * CHUNK_MB
            if allocated % 32 == 0 or (limit and allocated * 2**20 > limit * 0.8):
                events = cgroup.memory_events() or {}
                print(f"    allocated {allocated:5d} MiB   RSS {rss_mb():6.0f} MiB"
                      f"   oom_kill={events.get('oom_kill', 'n/a')}"
                      f"  high={events.get('high', 'n/a')}", flush=True)
            time.sleep(0.01)

    except MemoryError:
        # This runs only if you hit an RLIMIT_AS or the allocator genuinely
        # returned NULL. A cgroup OOM kill NEVER produces this. If you see
        # this traceback in a container, you were not OOM-killed -- you hit
        # a different limit, and the fix is a different fix.
        print()
        print("  MemoryError CAUGHT.", flush=True)
        print("  Read this carefully: a cgroup OOM kill cannot produce this exception.")
        print("  You hit an RLIMIT_AS, or the allocator returned NULL for another")
        print("  reason. Different mechanism, different fix. Check `ulimit -v`.")
        raise SystemExit(1)
    finally:
        # A finally block. Runs on exceptions and on SystemExit. Does not
        # run on SIGKILL.
        print(f"  [finally] reached, with {len(blocks) * CHUNK_MB} MiB allocated",
              flush=True)

    print()
    print(f"  Reached {len(blocks) * CHUNK_MB} MiB without being killed.")
    if limit is None:
        print("  Expected: there is no cgroup on this host to kill anything. The")
        print("  self-imposed ceiling stopped the loop; nothing was enforced.")
    else:
        print("  NOT expected under a memory limit. The kernel reclaimed enough to")
        print("  keep up -- allocate faster, or check that memory.max is what you")
        print("  think it is. If memory.high is set below max, that is the answer:")
        print("  the high counter above should be climbing, and this is the")
        print("  degrade-instead-of-die behaviour working exactly as advertised.")

    if args.free:
        print()
        print("  --free: dropping every reference and asking for a full GC.")
        before = rss_mb()
        blocks.clear()
        import gc

        gc.collect()
        time.sleep(0.5)
        after = rss_mb()
        if not HAVE_CURRENT_RSS:
            print(f"    peak RSS {after:.0f} MiB -- and PEAK is all this platform")
            print("    offers. There is no /proc/self/status here, so there is no")
            print("    current-RSS reading to compare against, and a peak that did")
            print("    not fall proves nothing. Run this in a Linux container to see")
            print("    the real before/after.")
        else:
            print(f"    RSS before {before:.0f} MiB -> after {after:.0f} MiB "
                  f"(returned {before - after:.0f} MiB)")
        print("    Freed objects go back to pymalloc's pools and arenas, and an")
        print("    arena returns to the OS only when it is COMPLETELY empty. So RSS")
        print("    is sticky: a burst that briefly needed this much leaves a process")
        print("    that looks like it needs this much, and your memory limit has to")
        print("    be sized for the peak rather than the average.")

    print()
    print("  Where the evidence lives when this DOES get killed:")
    print("    docker inspect <c> --format '{{.State.OOMKilled}}'   -> true")
    print("    exit code                                           -> 137 (128 + SIGKILL)")
    print("    cat /sys/fs/cgroup/memory.events                    -> oom_kill incremented")
    print("    dmesg on the host                                   -> the kill decision")
    print("    your application logs                               -> nothing at all")
    if events_before:
        print()
        print(f"  memory.events at start: {events_before}")
        print(f"  memory.events now     : {cgroup.memory_events()}")


if __name__ == "__main__":
    main()

File: 01-machine/07-inside-a-container/06-memory-the-limit-that-kills-you-without-a-traceback/cpp/oom.cpp
// 7.6 -- C++: the sharpest illustration on this page.
//
// WHAT THIS DEMONSTRATES
//   `new` is specified to throw std::bad_alloc on failure, so the obvious
//   defensive code is a try/catch around the allocation. This program writes
//   exactly that code -- and then prints, at the end, how many times the
//   catch block ran.
//
//   Under Linux's default overcommit it never runs. malloc returns a valid
//   pointer, the kernel commits nothing, and the cgroup charge lands when
//   you first WRITE to the page. So the failure arrives on a memory store,
//   arbitrarily far from the allocation, delivered as SIGKILL -- and you are
//   dead inside a memcpy, not inside your catch.
//
//   The language HAS the error path. The kernel's policy means you never
//   reach it. Reading that empty catch block is the most direct way to
//   internalise "the limit that kills you without a traceback", and it is
//   why C++ is the version worth writing this experiment in even though
//   nothing about it is C++-specific.
//
//   To make the contrast concrete, the program separates the two moments:
//
//     --reserve-only   allocate 4x the container limit and DO NOT touch it.
//                      Watch it succeed. Nothing is charged; nothing dies.
//     (default)        allocate and touch. Watch it die on the touch.
//
//   That pair is the entire difference between virtual address space and
//   resident memory, and it is the reason `top`'s VIRT column has misled a
//   generation of engineers.
//
//   (`vm.overcommit_memory=2` changes the policy and is worth knowing
//   exists. It is host-wide, with host-wide consequences, and it is not
//   something you set to fix one container.)
//
// WHAT TO LOOK FOR IN THE OUTPUT
//   1. In --reserve-only mode: gigabytes "allocated" with RSS flat. The
//      allocation genuinely succeeded and cost nothing.
//   2. In the default mode: the last line before the process disappears,
//      and the fact that "bad_alloc caught: 0" is never printed at all,
//      because the process does not survive to print its own summary.
//   3. The atexit hook and the destructor. Both will run on a normal exit.
//      Neither runs on SIGKILL.
//
// PORTABILITY
//   Builds and runs on Linux and macOS. Darwin has no cgroup memory
//   controller, no memory.events and no cgroup OOM killer, so with no limit
//   to read the program imposes its own ceiling and says clearly that it
//   stopped itself rather than being killed. The RSS reading uses
//   /proc/self/status on Linux and mach's task_info on Darwin.
//
// RUN
//   g++ -O2 -std=c++17 -o /tmp/oom oom.cpp && /tmp/oom
//
//   docker run --rm --memory=256m -v "$PWD:/w" -w /w gcc:14 \
//     sh -c 'g++ -O2 -std=c++17 -o /tmp/oom /w/oom.cpp && /tmp/oom'
//   echo "exit code: $?"      # 137
//   docker run --rm --memory=256m -v "$PWD:/w" -w /w gcc:14 \
//     sh -c 'g++ -O2 -std=c++17 -o /tmp/oom /w/oom.cpp && /tmp/oom --reserve-only'

#include <csignal>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <new>
#include <string>
#include <vector>

#if defined(__APPLE__)
#include <mach/mach.h>
#endif

namespace {

constexpr size_t kChunkMB = 8;

// ---------------------------------------------------------------- the kernel

// Bytes the cgroup will let this container charge, or 0 for no limit.
long long memory_max() {
  std::ifstream v2("/sys/fs/cgroup/memory.max");
  if (v2) {
    std::string raw;
    v2 >> raw;
    if (raw != "max" && !raw.empty()) return std::stoll(raw);
    return 0;
  }
  std::ifstream v1("/sys/fs/cgroup/memory/memory.limit_in_bytes");
  long long value = 0;
  // v1 spells "unlimited" as a number near 2^63, not as a word.
  if (v1 >> value && value < (1LL << 62)) return value;
  return 0;
}

std::string read_or(const char* path, const char* fallback) {
  std::ifstream file(path);
  if (!file) return fallback;
  std::string all, line;
  while (std::getline(file, line)) {
    if (!all.empty()) all += " ";
    all += line;
  }
  return all.empty() ? fallback : all;
}

// Current RSS in MiB. /proc does not exist on Darwin, so the two branches
// use genuinely different mechanisms rather than one that quietly returns a
// wrong number on the second platform.
double rss_mb() {
#if defined(__linux__)
  std::ifstream status("/proc/self/status");
  std::string line;
  while (std::getline(status, line)) {
    if (line.rfind("VmRSS:", 0) == 0)
      return std::stod(line.substr(6)) / 1024.0;
  }
  return -1.0;
#elif defined(__APPLE__)
  mach_task_basic_info info;
  mach_msg_type_number_t count = MACH_TASK_BASIC_INFO_COUNT;
  if (task_info(mach_task_self(), MACH_TASK_BASIC_INFO,
                reinterpret_cast<task_info_t>(&info), &count) != KERN_SUCCESS)
    return -1.0;
  return static_cast<double>(info.resident_size) / (1024.0 * 1024.0);
#else
  return -1.0;
#endif
}

int g_bad_alloc_caught = 0;

// A destructor, so you can watch it not run.
struct Farewell {
  ~Farewell() {
    std::cout << "  [destructor] returning from main normally -- RSS "
              << std::fixed << std::setprecision(0) << rss_mb() << " MiB\n";
  }
};

void at_exit_hook() {
  std::cout << "  [atexit] process exiting normally. bad_alloc caught: "
            << g_bad_alloc_caught << " time(s)\n";
}

// A signal handler that logs. It will run for SIGTERM. It will never run for
// SIGKILL, and that one sentence is the whole of this sub-topic.
void on_signal(int signum) {
  std::cout << "  [signal handler] caught signal " << signum
            << " -- shutting down cleanly\n";
  std::exit(128 + signum);
}

}  // namespace

int main(int argc, char** argv) {
  Farewell farewell;
  std::atexit(at_exit_hook);
  std::signal(SIGTERM, on_signal);
  std::signal(SIGINT, on_signal);

  bool reserve_only = false;
  size_t self_limit_mb = 512;
  for (int i = 1; i < argc; ++i) {
    std::string arg = argv[i];
    if (arg == "--reserve-only") reserve_only = true;
    if (arg == "--limit-mb" && i + 1 < argc) self_limit_mb = std::stoul(argv[++i]);
  }

  long long limit = memory_max();

  std::cout << "7.6 -- memory: C++\n";
#if defined(__linux__)
  std::cout << "  platform     : Linux\n";
#elif defined(__APPLE__)
  std::cout << "  platform     : Darwin (macOS) -- no cgroup memory controller\n";
#endif
  std::cout << "  memory.max   : "
            << (limit == 0 ? std::string("no limit / no cgroupfs")
                           : std::to_string(limit / (1 << 20)) + " MiB")
            << "\n";
  std::cout << "  memory.high  : " << read_or("/sys/fs/cgroup/memory.high", "unset")
            << "   <- degrades instead of killing; no Compose key\n";
  std::cout << "  heap ceiling : none. new goes to the system allocator and that is all.\n";
  std::cout << "  starting RSS : " << std::fixed << std::setprecision(0) << rss_mb()
            << " MiB\n\n";

  std::cout << "  installed, and about to be shown useless:\n";
  std::cout << "    * try/catch (std::bad_alloc) around every allocation\n";
  std::cout << "    * a SIGTERM handler that logs\n";
  std::cout << "    * an atexit hook\n";
  std::cout << "    * a destructor on a stack object in main\n\n";

  size_t ceiling_mb;
  if (limit == 0) {
    std::cout << "  !! No cgroup memory limit on this host, so nothing can OOM-kill\n";
    std::cout << "  !! this process. It will stop ITSELF at " << self_limit_mb
              << " MiB and say so.\n";
    std::cout << "  !! For the kill:\n";
    std::cout << "  !!   docker run --rm --memory=256m -v \"$PWD:/w\" -w /w gcc:14 \\\n";
    std::cout << "  !!     sh -c 'g++ -O2 -std=c++17 -o /tmp/oom /w/oom.cpp && /tmp/oom'\n\n";
    ceiling_mb = self_limit_mb;
  } else {
    ceiling_mb = static_cast<size_t>(limit / (1 << 20) * (reserve_only ? 4 : 1.5));
    std::cout << "  Allocating toward " << ceiling_mb << " MiB against a "
              << (limit / (1 << 20)) << " MiB limit.\n\n";
  }

  if (reserve_only) {
    std::cout << "  mode: --reserve-only. Allocating and NOT touching.\n";
    std::cout << "  Under default overcommit the kernel commits nothing until a page\n";
    std::cout << "  is written, so this should sail past the limit with RSS flat.\n\n";
  } else {
    std::cout << "  mode: allocate AND touch. The touch is what gets charged, and\n";
    std::cout << "  therefore what gets you killed.\n\n";
  }

  std::vector<char*> blocks;
  size_t allocated_mb = 0;

  while (allocated_mb < ceiling_mb) {
    char* block = nullptr;
    try {
      // The defensive code every careful C++ engineer writes. Count how
      // many times the handler below actually runs.
      block = new char[kChunkMB << 20];
    } catch (const std::bad_alloc& err) {
      ++g_bad_alloc_caught;
      std::cout << "\n  bad_alloc CAUGHT after " << allocated_mb << " MiB: "
                << err.what() << "\n";
      std::cout << "  The ALLOCATOR refused. That is a different event from a cgroup\n";
      std::cout << "  OOM kill, which never gives the allocator a chance to refuse.\n";
      std::cout << "  If you are in a container and reached this line, check for an\n";
      std::cout << "  RLIMIT_AS (ulimit -v) or vm.overcommit_memory=2.\n";
      break;
    }

    if (!reserve_only) {
      // Touch every page. One byte per 4 KiB is enough -- the charge is per
      // page, not per byte. THIS is the line the kernel kills you on.
      for (size_t offset = 0; offset < (kChunkMB << 20); offset += 4096)
        block[offset] = 1;
    }
    blocks.push_back(block);
    allocated_mb += kChunkMB;

    if (allocated_mb % 64 == 0 ||
        (limit != 0 && static_cast<long long>(allocated_mb) << 20 > limit * 8 / 10)) {
      std::cout << "    allocated " << std::setw(5) << allocated_mb << " MiB   RSS "
                << std::setw(6) << std::fixed << std::setprecision(0) << rss_mb()
                << " MiB   memory.events: "
                << read_or("/sys/fs/cgroup/memory.events", "n/a") << std::endl;
    }
  }

  std::cout << "\n  Reached " << allocated_mb << " MiB. bad_alloc caught "
            << g_bad_alloc_caught << " time(s).\n";

  if (reserve_only) {
    std::cout << "  Note the RSS column against the allocated column. The gap between\n";
    std::cout << "  them is memory that exists in your address space and nowhere else.\n";
    std::cout << "  That gap is why `top`'s VIRT column has misled a generation of\n";
    std::cout << "  engineers, and why an allocation succeeding tells you nothing\n";
    std::cout << "  about whether you have the memory.\n";
  } else if (limit == 0) {
    std::cout << "  Expected: no cgroup here to kill anything, and the self-imposed\n";
    std::cout << "  ceiling stopped the loop. Nothing was enforced.\n";
  } else {
    std::cout << "  NOT expected under a memory limit while touching every page. The\n";
    std::cout << "  kernel reclaimed enough to keep up, or memory.high is set.\n";
  }

  std::cout << "\n  If bad_alloc was caught 0 times and the process is still alive, the\n";
  std::cout << "  catch block above is decoration. In a container it is decoration\n";
  std::cout << "  ALWAYS: under default overcommit the allocator does not fail, the\n";
  std::cout << "  kernel kills you on a store instruction, and SIGKILL cannot be\n";
  std::cout << "  caught, blocked or handled by anything -- including everything this\n";
  std::cout << "  program installed at the top.\n";

  for (char* block : blocks) delete[] block;
  return 0;
}
